[doc] add sqlite/next doc (#24962)
expo/expo: docs/pages/versions/unversioned/sdk/sqlite-next.mdx

---
title: SQLite (Next)
description: A library that provides access to a database that can be queried through a SQLite API.
sourceCodeUrl: 'https://github.com/expo/expo/tree/main/packages/expo-sqlite'
packageName: 'expo-sqlite'
---

import APISection from '~/components/plugins/APISection';
import { APIInstallSection } from '~/components/plugins/InstallSection';
import PlatformsSection from '~/components/plugins/PlatformsSection';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';
import { BoxLink } from '~/ui/components/BoxLink';
import { GithubIcon } from '@expo/styleguide-icons';

`expo-sqlite` gives your app access to a database that can be queried through a SQLite API. The database is persisted across restarts of your app.

<PlatformsSection android emulator ios simulator />

## Installation

<APIInstallSection />

## Usage

> ### Importing an existing database

Reviewer: ```suggestion
### Import an existing database
```
Let's use short form heading titles instead of gerunds.

> To open a new SQLite database using an existing `.db` file you already have, follow the steps below:

Reviewer: ```suggestion
To open a new SQLite database using an existing **.db** file you already have, follow the steps below:
```

[Filename as bold text](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#filenames-as-bold-text) as per our writing style guide

<Step label="1">

> Install `expo-file-system` and `expo-asset` modules:

Reviewer: ```suggestion
Install `expo-file-system` and `expo-asset` libraries:
```

[As per our glossary in writing style guide and page description, let's use library/libraries instead of modules here](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#glossary)

<Terminal cmd={['$ npx expo install expo-file-system expo-asset']} />

</Step>

<Step label="2">
Create a **metro.config.js** file at the root of your project with the following contents to include [extra asset extensions](/guides/customizing-metro/#adding-more-file-extensions-to--assetexts):

> ```js

Reviewer: ````suggestion
```js metro.config.js
````

  'INSERT INTO test (value, intValue) VALUES ($value, $intValue)'
);
await statement.runAsync({ $value: 'bbb', $intValue: 101 });
await statement.runAsync({ $value: 'ccc', $intValue: 102 });
await statement.runAsync({ $value: 'ddd', $intValue: 103 });
await statement.finalizeAsync();

// `getAsync` is useful when you want to get a single row from the database.
const oneRow = await db.getAsync('SELECT * FROM test');
console.log(oneRow.id, oneRow.value, oneRow.intValue);

// `allAsync` is useful when you want to get all results as an array of objects.
const allRows = await db.allAsync('SELECT * FROM test');
for (const row of allRows) {
  console.log(row.id, row.value, row.intValue);
}

// `eachAsync` is useful when you want to iterate SQLite query cursor.
for await (const row of db.eachAsync('SELECT * FROM test')) {
  console.log(row.id, row.value, row.intValue);
}
```

### Executing queries within an async transaction

```js
const db = await SQLite.openDatabaseAsync('dbName');

await db.transactionAsync(async () => {
  const result = await db.getAsync('SELECT COUNT(*) FROM USERS');
  console.log('Count:', result.rows[0]['COUNT(*)']);
});
```

> > **Note:** Please note that due to the nature of the async/await, We can't ensure that the transaction will be executed in the same order as the statements are written. A query out of a transaction may be executed inside the transaction. If you care about the order of execution, you should use the [`transactionExclusiveAsync`](#transactionexclusiveasynctask) function.

Reviewer: ```suggestion
> **Note:** Due to the nature of the async/await, we can't ensure that the transaction will be executed in the same order as the statements are written. A query out of a transaction may be executed inside the transaction. If you care about the order of execution, use the [`transactionExclusiveAsync`](#transactionexclusiveasynctask) function.
```

- We don't have to use words like please. It's okay to be a bit authoritative.
- Fix typo

### Executing PRAGMA queries

```js
const db = await SQLite.openDatabaseAsync('dbName');
await db.execAsync('PRAGMA journal_mode = WAL');
await db.execAsync('PRAGMA foreign_keys = ON');
```

> > **Tip:** Enable WAL journal mode when you create a new database to improve performance in general. [Learn more](https://www.sqlite.org/wal.html).

Reviewer: ```suggestion
> **info** **Tip:** Enable [WAL journal mode](https://www.sqlite.org/wal.html) when you create a new database to improve performance in general.
```

- Let's use info type of callout to provide the tip here.
- Also, instead of adding learn more as context to provide link, we can just provide the link for the actual context.
